File: Repositorio/engenharia-de-software-com-ia-aplicada-main/modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-05-avaliacao-modelos/npv-real-vs-projetado-tool.js
/**
 * Ahirton Lopes · Fine-Tuning Toolkit
 * Artefato de Demo - Módulo 5.4 (companion, fecha o arco financeiro do Módulo 1)
 *
 * O Módulo 1.3 rodou a projeção de NPV que aprovou Amplitude Auto
 * (Saúde Empresarial só foi aprovada depois, no Módulo 3.2) com um custo
 * de treino ESTIMADO (R$2.400, o valor de referência desta disciplina pra "GPU
 * alugada"). O Módulo 5.2 mediu o custo REAL desses treinos no billing
 * do Google Cloud: R$1,53 (Auto) e R$0,86 (Saúde Empresarial) -- centavos,
 * não milhares de reais.
 *
 * Este arquivo NÃO inventa nenhum dado de negócio novo (nenhuma receita
 * de produção, nenhum ROI fictício): reabre o MESMO código de NPV do
 * Módulo 1.3 (decision-framework-tool.js, `calcularNPV` e
 * `simularMonteCarlo`, sem duplicar nenhuma lógica financeira) e troca
 * só o custo de treino, do estimado pro medido, mantendo toda a outra
 * premissa (crescimento de volume, custo por chamada) exatamente como a
 * projeção original -- essas continuam sendo estimativa, não foram
 * medidas em produção.
 *
 * Uso: node npv-real-vs-projetado-tool.js
 *
 * Par oficial desta disciplina: npv-real-vs-projetado-tool.js (oficial) /
 * npv_real_vs_projetado_tool.py (referência espelhada).
 */

'use strict';

const assert = require('assert').strict;
const path = require('path');

const {
  carregarConfiguracao,
  calcularNPV,
  simularMonteCarlo,
} = require(path.join(__dirname, '..', 'modulo-01-decision-framework', 'decision-framework-tool.js'));

/* ============================================================================
 * 1. Custo de treino real, medido no Módulo 5.2 (billing real do Google
 *    Cloud), contra o valor de referência estimado no Módulo 1.3
 * ========================================================================= */

const CUSTO_TREINAMENTO_REAL = Object.freeze({
  'amplitude-auto': 1.53,
  'amplitude-saude-empresarial': 0.86,
});

/* ============================================================================
 * 2. Montar os params determinísticos que calcularNPV espera (mesma forma
 *    que a função interna paramsDeterministicos do Módulo 1.3, não
 *    exportada -- reconstruída aqui só como adaptador de interface, a
 *    lógica financeira em si continua vindo de calcularNPV real)
 * ========================================================================= */

function montarParamsDeterministicos(financeiro, custoTreinamentoOverride = null) {
  return {
    volumeInicialMensal: financeiro.volumeInicialMensal,
    crescimentoMensal: financeiro.crescimentoMensal.moda,
    custoPorChamadaStatusQuo: financeiro.custoPorChamadaStatusQuo.moda,
    custoPorChamadaFineTuned: financeiro.custoPorChamadaFineTuned.moda,
    custoTreinamento: custoTreinamentoOverride !== null ? custoTreinamentoOverride : financeiro.custoTreinamento,
    horizonteMeses: financeiro.horizonteMeses,
    taxaDescontoMensal: financeiro.taxaDescontoMensal,
  };
}

/* ============================================================================
 * 3. Comparar projeção original (custo estimado) contra real (custo medido)
 * ========================================================================= */

function compararProjetadoVsReal(casoId, config) {
  const caso = config.casos.find((c) => c.id === casoId);
  if (!caso) throw new Error(`caso não encontrado: ${casoId}`);
  const custoReal = CUSTO_TREINAMENTO_REAL[casoId];
  if (custoReal === undefined) throw new Error(`sem custo real medido pra: ${casoId}`);

  const projetado = calcularNPV(montarParamsDeterministicos(caso.financeiro));
  const real = calcularNPV(montarParamsDeterministicos(caso.financeiro, custoReal));

  return {
    casoId,
    custoTreinamentoProjetado: caso.financeiro.custoTreinamento,
    custoTreinamentoReal: custoReal,
    npvProjetado: projetado.npv,
    breakevenProjetado: projetado.mesBreakeven,
    npvReal: real.npv,
    breakevenReal: real.mesBreakeven,
  };
}

/* --------------------------------------------------------------------------
 * Testes automatizados -- contra os números reais capturados rodando este
 * mesmo código nesta máquina em 2026-09-04
 * -------------------------------------------------------------------------- */

let totalTestes = 0;
let testesComFalha = 0;

function testar(descricao, fn) {
  totalTestes += 1;
  try {
    fn();
    console.log(`  [OK] ${descricao}`);
  } catch (erro) {
    testesComFalha += 1;
    console.log(`  [FALHOU] ${descricao}`);
    console.log(`           ${erro.message}`);
  }
}

function rodarTestes() {
  console.log('== Testes: recomputação de NPV com custo de treino real ==');
  const config = carregarConfiguracao();

  testar('Auto: NPV projetado bate com o número real do Módulo 1.3 (R$4.780,27, breakeven mês 10)', () => {
    const r = compararProjetadoVsReal('amplitude-auto', config);
    assert.equal(r.custoTreinamentoProjetado, 2400);
    assert.equal(r.npvProjetado, 4780.27);
    assert.equal(r.breakevenProjetado, 10);
  });

  testar('Auto: NPV real (custo R$1,53) sobe pra R$7.178,74, breakeven cai pro mês 1', () => {
    const r = compararProjetadoVsReal('amplitude-auto', config);
    assert.equal(r.npvReal, 7178.74);
    assert.equal(r.breakevenReal, 1);
    assert.ok(r.npvReal > r.npvProjetado, 'NPV real deveria ser maior que o projetado');
  });

  testar('Saúde Empresarial: NPV projetado era negativo (R$-993,23, sem breakeven)', () => {
    const r = compararProjetadoVsReal('amplitude-saude-empresarial', config);
    assert.equal(r.custoTreinamentoProjetado, 2400);
    assert.equal(r.npvProjetado, -993.23);
    assert.equal(r.breakevenProjetado, null);
  });

  testar('Saúde Empresarial: NPV real (custo R$0,86) vira positivo, R$1.405,91, breakeven mês 1', () => {
    const r = compararProjetadoVsReal('amplitude-saude-empresarial', config);
    assert.equal(r.npvReal, 1405.91);
    assert.equal(r.breakevenReal, 1);
  });

  testar('custo por chamada e crescimento de volume NÃO mudam (continuam projeção, não medidos)', () => {
    const caso = config.casos.find((c) => c.id === 'amplitude-auto');
    const paramsProjetado = montarParamsDeterministicos(caso.financeiro);
    const paramsReal = montarParamsDeterministicos(caso.financeiro, CUSTO_TREINAMENTO_REAL['amplitude-auto']);
    assert.equal(paramsProjetado.custoPorChamadaFineTuned, paramsReal.custoPorChamadaFineTuned);
    assert.equal(paramsProjetado.crescimentoMensal, paramsReal.crescimentoMensal);
    assert.notEqual(paramsProjetado.custoTreinamento, paramsReal.custoTreinamento);
  });

  console.log(`\n${totalTestes - testesComFalha}/${totalTestes} testes passaram.`);
  if (testesComFalha > 0) process.exitCode = 1;
}

/* ============================================================================
 * 4. Comparação real, impressa
 * ========================================================================= */

function rodarComparacaoReal() {
  const config = carregarConfiguracao();
  console.log('\n===== NPV: projeção original (Módulo 1.3) vs. custo real medido (Módulo 5.2) =====\n');

  ['amplitude-auto', 'amplitude-saude-empresarial'].forEach((casoId) => {
    const r = compararProjetadoVsReal(casoId, config);
    console.log(`--- ${casoId} ---`);
    console.log(`  Custo de treino projetado: R$${r.custoTreinamentoProjetado.toFixed(2)}  ->  real medido: R$${r.custoTreinamentoReal.toFixed(2)}`);
    console.log(`  NPV projetado: R$${r.npvProjetado.toFixed(2)} (breakeven: ${r.breakevenProjetado === null ? 'nunca' : `mês ${r.breakevenProjetado}`})`);
    console.log(`  NPV real:      R$${r.npvReal.toFixed(2)} (breakeven: ${r.breakevenReal === null ? 'nunca' : `mês ${r.breakevenReal}`})`);
    console.log();
  });

  console.log(
    'O que mudou: só o custo de treino, de estimativa pra medição real. Custo por '
    + 'chamada em produção e crescimento de volume continuam sendo a mesma projeção '
    + 'do Módulo 1.3 -- não foram medidos em produção, então não mudam aqui.'
  );

  return ['amplitude-auto', 'amplitude-saude-empresarial'].map((id) => compararProjetadoVsReal(id, config));
}

if (require.main === module) {
  rodarTestes();
  rodarComparacaoReal();
}

module.exports = {
  CUSTO_TREINAMENTO_REAL,
  montarParamsDeterministicos,
  compararProjetadoVsReal,
  rodarComparacaoReal,
};

/*
 * Ahirton Lopes · Fine-Tuning Toolkit - UNIPDS: Processamento de Dados e Fine-Tuning de Modelos
 * Prof. Ahirton Lopes, Ph.D. - GDE AI, Microsoft MVP, Senior Manager
 */

File: Repositorio/engenharia-de-software-com-ia-aplicada-main/modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-05-avaliacao-modelos/npv_real_vs_projetado_tool.py
"""
Ahirton Lopes - Fine-Tuning Toolkit
Artefato de Demo - Modulo 5.4 (companion, fecha o arco financeiro do Modulo 1,
referencia espelhada em Python do .js oficial)

O Modulo 1.3 rodou a projecao de NPV que aprovou Amplitude Auto
(Saude Empresarial so foi aprovada depois, no Modulo 3.2) com um custo
de treino ESTIMADO (R$2.400, o
valor de referencia desta disciplina pra "GPU alugada"). O Modulo 5.2 mediu
o custo REAL desses treinos no billing do Google Cloud: R$1,53 (Auto) e
R$0,86 (Saude Empresarial) -- centavos, nao milhares de reais.

Este arquivo NAO inventa nenhum dado de negocio novo (nenhuma receita de
producao, nenhum ROI ficticio): reabre o MESMO codigo de NPV do Modulo 1.3
(decision_framework_tool.py, calcular_npv e simular_monte_carlo, sem
duplicar nenhuma logica financeira) e troca so o custo de treino, do
estimado pro medido, mantendo toda a outra premissa (crescimento de
volume, custo por chamada) exatamente como a projecao original -- essas
continuam sendo estimativa, nao foram medidas em producao.

Uso: python3 npv_real_vs_projetado_tool.py
"""

import importlib.util
from pathlib import Path

_M1_2_PATH = Path(__file__).parent.parent / "modulo-01-decision-framework" / "decision_framework_tool.py"
_spec = importlib.util.spec_from_file_location("decision_framework_tool", _M1_2_PATH)
dft = importlib.util.module_from_spec(_spec)
_spec.loader.exec_module(dft)


# =============================================================================
# 1. Custo de treino real, medido no Modulo 5.2 (billing real do Google
#    Cloud), contra o valor de referencia estimado no Modulo 1.3
# =============================================================================

CUSTO_TREINAMENTO_REAL = {
    "amplitude-auto": 1.53,
    "amplitude-saude-empresarial": 0.86,
}


# =============================================================================
# 2. Montar os params deterministicos que calcular_npv espera (mesma forma
#    que a funcao interna _params_deterministicos do Modulo 1.3 -- adaptador
#    de interface, a logica financeira em si continua vindo de calcular_npv
#    real)
# =============================================================================

def montar_params_deterministicos(financeiro, custo_treinamento_override=None):
    # As chaves seguem o JSON de configuração compartilhado com o .js oficial
    # (amplitude-seguros-casos.json), que é camelCase -- calcular_npv espera
    # exatamente essas chaves, mesmo no lado Python.
    return {
        "volumeInicialMensal": financeiro["volumeInicialMensal"],
        "crescimentoMensal": financeiro["crescimentoMensal"]["moda"],
        "custoPorChamadaStatusQuo": financeiro["custoPorChamadaStatusQuo"]["moda"],
        "custoPorChamadaFineTuned": financeiro["custoPorChamadaFineTuned"]["moda"],
        "custoTreinamento": (
            custo_treinamento_override if custo_treinamento_override is not None
            else financeiro["custoTreinamento"]
        ),
        "horizonteMeses": financeiro["horizonteMeses"],
        "taxaDescontoMensal": financeiro["taxaDescontoMensal"],
    }


# =============================================================================
# 3. Comparar projecao original (custo estimado) contra real (custo medido)
# =============================================================================

def comparar_projetado_vs_real(caso_id, config):
    caso = next((c for c in config["casos"] if c["id"] == caso_id), None)
    if caso is None:
        raise ValueError(f"caso nao encontrado: {caso_id}")
    custo_real = CUSTO_TREINAMENTO_REAL.get(caso_id)
    if custo_real is None:
        raise ValueError(f"sem custo real medido pra: {caso_id}")

    projetado = dft.calcular_npv(montar_params_deterministicos(caso["financeiro"]))
    real = dft.calcular_npv(montar_params_deterministicos(caso["financeiro"], custo_real))

    return {
        "caso_id": caso_id,
        "custo_treinamento_projetado": caso["financeiro"]["custoTreinamento"],
        "custo_treinamento_real": custo_real,
        "npv_projetado": projetado["npv"],
        "breakeven_projetado": projetado["mesBreakeven"],
        "npv_real": real["npv"],
        "breakeven_real": real["mesBreakeven"],
    }


# -----------------------------------------------------------------------------
# Testes automatizados -- contra os numeros reais capturados rodando este
# mesmo codigo nesta maquina em 2026-09-04
# -----------------------------------------------------------------------------

_total_testes = 0
_testes_com_falha = 0


def testar(descricao, fn):
    global _total_testes, _testes_com_falha
    _total_testes += 1
    try:
        fn()
        print(f"  [OK] {descricao}")
    except Exception as erro:  # noqa: BLE001
        _testes_com_falha += 1
        print(f"  [FALHOU] {descricao}")
        print(f"           {erro}")


def rodar_testes():
    print("== Testes: recomputacao de NPV com custo de treino real ==")
    config = dft.carregar_configuracao()

    def teste_auto_projetado():
        r = comparar_projetado_vs_real("amplitude-auto", config)
        assert r["custo_treinamento_projetado"] == 2400
        assert r["npv_projetado"] == 4780.27
        assert r["breakeven_projetado"] == 10
    testar("Auto: NPV projetado bate com o numero real do Modulo 1.3 (R$4.780,27, breakeven mes 10)", teste_auto_projetado)

    def teste_auto_real():
        r = comparar_projetado_vs_real("amplitude-auto", config)
        assert r["npv_real"] == 7178.74
        assert r["breakeven_real"] == 1
        assert r["npv_real"] > r["npv_projetado"]
    testar("Auto: NPV real (custo R$1,53) sobe pra R$7.178,74, breakeven cai pro mes 1", teste_auto_real)

    def teste_saude_projetado():
        r = comparar_projetado_vs_real("amplitude-saude-empresarial", config)
        assert r["custo_treinamento_projetado"] == 2400
        assert r["npv_projetado"] == -993.23
        assert r["breakeven_projetado"] is None
    testar("Saude Empresarial: NPV projetado era negativo (R$-993,23, sem breakeven)", teste_saude_projetado)

    def teste_saude_real():
        r = comparar_projetado_vs_real("amplitude-saude-empresarial", config)
        assert r["npv_real"] == 1405.91
        assert r["breakeven_real"] == 1
    testar("Saude Empresarial: NPV real (custo R$0,86) vira positivo, R$1.405,91, breakeven mes 1", teste_saude_real)

    def teste_outros_params_intocados():
        caso = next(c for c in config["casos"] if c["id"] == "amplitude-auto")
        p_projetado = montar_params_deterministicos(caso["financeiro"])
        p_real = montar_params_deterministicos(caso["financeiro"], CUSTO_TREINAMENTO_REAL["amplitude-auto"])
        assert p_projetado["custoPorChamadaFineTuned"] == p_real["custoPorChamadaFineTuned"]
        assert p_projetado["crescimentoMensal"] == p_real["crescimentoMensal"]
        assert p_projetado["custoTreinamento"] != p_real["custoTreinamento"]
    testar("custo por chamada e crescimento de volume NAO mudam (continuam projecao, nao medidos)", teste_outros_params_intocados)

    print(f"\n{_total_testes - _testes_com_falha}/{_total_testes} testes passaram.")
    return _testes_com_falha == 0


# =============================================================================
# 4. Comparacao real, impressa
# =============================================================================

def rodar_comparacao_real():
    config = dft.carregar_configuracao()
    print("\n===== NPV: projecao original (Modulo 1.3) vs. custo real medido (Modulo 5.2) =====\n")

    resultados = []
    for caso_id in ["amplitude-auto", "amplitude-saude-empresarial"]:
        r = comparar_projetado_vs_real(caso_id, config)
        resultados.append(r)
        print(f"--- {caso_id} ---")
        print(f"  Custo de treino projetado: R${r['custo_treinamento_projetado']:.2f}  ->  real medido: R${r['custo_treinamento_real']:.2f}")
        be_p = "nunca" if r["breakeven_projetado"] is None else f"mes {r['breakeven_projetado']}"
        be_r = "nunca" if r["breakeven_real"] is None else f"mes {r['breakeven_real']}"
        print(f"  NPV projetado: R${r['npv_projetado']:.2f} (breakeven: {be_p})")
        print(f"  NPV real:      R${r['npv_real']:.2f} (breakeven: {be_r})")
        print()

    print(
        "O que mudou: so o custo de treino, de estimativa pra medicao real. Custo por "
        "chamada em producao e crescimento de volume continuam sendo a mesma projecao "
        "do Modulo 1.3 -- nao foram medidos em producao, entao nao mudam aqui."
    )

    return resultados


if __name__ == "__main__":
    ok = rodar_testes()
    rodar_comparacao_real()
    if not ok:
        raise SystemExit(1)

# Ahirton Lopes - Fine-Tuning Toolkit - UNIPDS: Processamento de Dados e Fine-Tuning de Modelos
# Prof. Ahirton Lopes, Ph.D. - GDE AI, Microsoft MVP, Senior Manager
